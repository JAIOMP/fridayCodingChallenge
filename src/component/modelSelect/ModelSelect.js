import React, { useEffect, useContext, useState } from 'react';
import SelectItem from '../common/selectItem/SelectItem'
import VehicleContext from "../../provider/VehicleDataContext";
import './ModelSelect.css';

const ModelSelect = () => {
    const [models, setModels] = useState([]);
    const { selectedBrand, setSelectedModel, setError } = useContext(VehicleContext);

    useEffect(() => {
        setError('');
        fetch(`http://localhost:8080/api/models?make=${selectedBrand}`)
            .then(res => res.json())
            .then(data => {
                if(data.length === 0) {
                    setError('EMPTY_MODEL_DATA');
                }
                setModels(data)
            }).catch(() => {
                setError('API_ERROR');
        });
    }, [selectedBrand]);

    const onSelect = (event) => {
        setSelectedModel(event.target.value);
    };

    return (
        <>
            {models.length > 0  &&
            <SelectItem data={models} dataFor="model" onSelect={onSelect}/>}
        </>
    );
};

export default ModelSelect;
