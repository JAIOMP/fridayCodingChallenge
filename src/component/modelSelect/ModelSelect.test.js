import React from 'react';
import {render, waitFor} from '@testing-library/react';
import ModelSelect from './ModelSelect';
import VehicleContext from "../../provider/VehicleDataContext";

describe('ModelSelect', () => {
    const mockResponse = ["1er", "2er", "3er", "4er", "5er", "6er", "7er", "8er"];

    const contextData = {
        setError: jest.fn(),
        setSelectedModel: jest.fn(),
        selectedBrand: "Ford",
    };
    beforeEach(() => {
        jest.spyOn(global, 'fetch').mockResolvedValue({
            json: jest.fn().mockResolvedValue(mockResponse)
        })
    });

    it('should render ModelSelect', async () => {
        const {container, getByTestId} = render(
            <VehicleContext.Provider value={contextData}>
                <ModelSelect/>
            </VehicleContext.Provider>
        );

        await waitFor(() => getByTestId("select-item"));
        expect(container).toMatchSnapshot();
    });

    it('should render empty model data error when model data is empty', async () => {
        jest.spyOn(global, 'fetch').mockResolvedValue({
            json: jest.fn().mockResolvedValue([])
        });

        render(
            <VehicleContext.Provider value={contextData}>
                <ModelSelect/>
            </VehicleContext.Provider>
        );

        await waitFor(() => {
            expect(contextData.setError).toHaveBeenCalledTimes(2);
            expect(contextData.setError).toHaveBeenNthCalledWith(1, "");
            expect(contextData.setError).toHaveBeenNthCalledWith(2, "EMPTY_MODEL_DATA");
        })
    });

    it('should render API_ERROR when api fails', async () => {
        jest.spyOn(global, 'fetch').mockImplementation(() => Promise.reject("API is down"));

        render(
            <VehicleContext.Provider value={contextData}>
                <ModelSelect/>
            </VehicleContext.Provider>
        );

        await waitFor(() => {
            expect(contextData.setError).toHaveBeenCalledTimes(2);
            expect(contextData.setError).toHaveBeenNthCalledWith(1, "");
            expect(contextData.setError).toHaveBeenNthCalledWith(2, "API_ERROR");
        })
    });
});
