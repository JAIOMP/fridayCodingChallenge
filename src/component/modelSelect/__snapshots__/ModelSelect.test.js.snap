// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelSelect should render ModelSelect 1`] = `
<div>
  <div
    class="select-item-container"
  >
    <label
      class="select-label"
      data-testid="select-item"
      for="brand"
    >
      Please select your 
      <strong>
        model
      </strong>
      :
    </label>
    <select
      data-testid="select"
      id="model-list"
    >
      <option>
        --Please choose an option--
      </option>
      <option
        data-testid="select-option"
        value="1er"
      >
        1er
      </option>
      <option
        data-testid="select-option"
        value="2er"
      >
        2er
      </option>
      <option
        data-testid="select-option"
        value="3er"
      >
        3er
      </option>
      <option
        data-testid="select-option"
        value="4er"
      >
        4er
      </option>
      <option
        data-testid="select-option"
        value="5er"
      >
        5er
      </option>
      <option
        data-testid="select-option"
        value="6er"
      >
        6er
      </option>
      <option
        data-testid="select-option"
        value="7er"
      >
        7er
      </option>
      <option
        data-testid="select-option"
        value="8er"
      >
        8er
      </option>
    </select>
  </div>
</div>
`;
